src/persist-resource.test.js

```diff
@@ -0,0 +1,104 @@
+import React from "react";
+import { render, waitForElement } from "react-testing-library";
+import { identity } from "ramda";
+import useEntry from "./use-entry";
+import RemoteResourceBoundary from "./RemoteResourceBoundary";
+import persistResource from "./persist-resource";
+import createKeyedResource from "./create-keyed-resource";
+
+// ---------------------------
+// Mocks
+// ---------------------------
+
+const MockResourceConsumer = ({ resource, index }) => {
+  const [entry] = useEntry(resource, [index]);
+
+  return entry;
+};
+
+// ---------------------------
+// Tests
+// ---------------------------
+
+describe("persistResource", () => {
+  it("loads the initial state into the resource", async () => {
+    const getInitialState = jest
+      .fn()
+      .mockImplementation(() => Promise.resolve({ 0: "initial" }));
+    const persist = jest.fn();
+
+    const resource = persistResource(
+      getInitialState,
+      persist,
+      createKeyedResource(identity, () => Promise.resolve("load"))
+    );
+
+    const { getByText } = render(
+      <RemoteResourceBoundary
+        fallback={<p>Loading...</p>}
+        renderError={() => <p>error</p>}
+      >
+        <MockResourceConsumer resource={resource} index={0} />
+      </RemoteResourceBoundary>
+    );
+
+    await waitForElement(() => getByText("initial"));
+  });
+
+  it("runs the persist function when the state updates", async () => {
+    const getInitialState = jest
+      .fn()
+      .mockImplementation(() => Promise.resolve({ 0: "initial" }));
+    const persist = jest.fn();
+
+    const resource = persistResource(
+      getInitialState,
+      persist,
+      createKeyedResource(identity, () => Promise.resolve("load"))
+    );
+
+    const { getByText } = render(
+      <RemoteResourceBoundary
+        fallback={<p>Loading...</p>}
+        renderError={() => <p>error</p>}
+      >
+        <MockResourceConsumer resource={resource} index={0} />
+      </RemoteResourceBoundary>
+    );
+
+    await waitForElement(() => getByText("initial"));
+
+    resource.setState({ 0: "setState" });
+    await waitForElement(() => getByText("setState"));
+
+    expect(persist).toHaveBeenNthCalledWith(1, { 0: "setState" });
+    expect(persist).toHaveBeenCalledTimes(1);
```

Comment:
Does we need this assertion? Maybe I'm misunderstanding `toHaveBeenNthCalledWith`, but it appears that it already tests the amount of times it has been called.

Reply:
Got ya! `toHaveBeenNthCalledWith` checks to make sure the nth call to the spy had  specific arguments. The `toHaveBeenCalledTimes(1)` ensures the function is only called once.

Reply:
Ahhhh, gotcha. That makes sense!